modules/repository/prune_hook_task.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package repository
+
+import (
+	"context"
+	"fmt"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/log"
+
+	"xorm.io/builder"
+)
+
+// PruneHookTaskTable deletes rows from hook_task as needed.
+func PruneHookTaskTable(ctx context.Context) error {
+	log.Error("Doing: PruneHookTaskTable")
+
+	if err := models.Iterate(
+		models.DefaultDBContext(),
+		new(models.Repository),
+		builder.Expr("id>0 AND is_hook_task_purge_enabled=?", true),
+		func(idx int, bean interface{}) error {
+			select {
+			case <-ctx.Done():
+				return fmt.Errorf("Aborted due to shutdown")
+			default:
+			}
+			repo := bean.(*models.Repository)
+			repoPath := repo.RepoPath()
+			log.Trace("Running prune hook_task table on repository %s", repoPath)
+			if err := models.DeleteDeliveredHookTasks(repo.ID, repo.NumberWebhookDeliveriesToKeep); err != nil {
```

Comment:
This may fail on sqlite. I think use a `Find` but not `Iterate` is better.

Reply:
I'm not sure I totally understand, can you elaborate? I saw Iterate usage other places like for example (if I'm understanding):
https://github.com/go-gitea/gitea/blob/master/modules/repository/check.go#L25
https://github.com/go-gitea/gitea/blob/master/modules/repository/hooks.go#L159

But I'm happy to refactor it if it is a problem.

Reply:
When did some database operations on an iterate maybe affect a lock on sqlite. If other places also have the problems, they also needs to be changed.